charmhelpers/contrib/openstack/deferred_events.py

```diff
@@ -0,0 +1,291 @@
+# Copyright 2021 Canonical Limited.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#  http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Module for managing deferred service events.
+
+This module is used to manage deferred service events from both charm actions
+and package actions.
+"""
+
+import collections
+import datetime
+import time
+
+import charmhelpers.core as ch_core
+import charmhelpers.contrib.openstack.policy_rcd as policy_rcd
+import charmhelpers.core.hookenv as hookenv
+import charmhelpers.core.host as host
+
+import subprocess
+
+
+ServiceEvent = collections.namedtuple(
+    'ServiceEvent',
+    'timestamp service reason action')
+
+
+class DeferredServiceEvents():
+    """A class to manage deferred service events."""
+
+    def __init__(self):
+        """Intialise deferred events class."""
+        self.events = self.load_events()
+
+
+class DeferredCharmServiceEvents(DeferredServiceEvents):
+    """Deferred service events from charm actions."""
+
+    KV_KEY = 'deferred_events'
+
+    def load_events(self):
+        """Load events previously requested by the charm.
+
+        :returns: List of events requested by the charm.
+        :rtype: List[ServiceEvent]
+        """
+        with ch_core.unitdata.HookData()() as t:
+            kv = t[0]
+            queued_restarts = kv.get(self.KV_KEY, [])
```

Comment:
For generalization should this be queued_events?

Reply:
This class has now been removed